homeassistant/components/habitica/calendar.py

```diff
@@ -0,0 +1,217 @@
+"""Calendar platform for Habitica integration."""
+
+from __future__ import annotations
+
+from datetime import date, datetime, timedelta
+from enum import StrEnum
+
+from dateutil.rrule import rrule
+
+from homeassistant.components.calendar import (
+    CalendarEntity,
+    CalendarEntityDescription,
+    CalendarEvent,
+)
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.util import dt as dt_util
+
+from . import HabiticaConfigEntry
+from .coordinator import HabiticaDataUpdateCoordinator
+from .entity import HabiticaBase
+from .types import HabiticaTaskType
+from .util import build_rrule, get_recurrence_rule
+
+
+class HabiticaCalendar(StrEnum):
+    """Habitica calendars."""
+
+    DAILIES = "dailys"
+    TODOS = "todos"
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: HabiticaConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up the calendar platform."""
+    coordinator = config_entry.runtime_data
+
+    async_add_entities(
+        [
+            HabiticaTodosCalendarEntity(coordinator),
+            HabiticaDailiesCalendarEntity(coordinator),
+        ]
+    )
+
+
+class HabiticaCalendarEntity(HabiticaBase, CalendarEntity):
+    """Base Habitica calendar entity."""
+
+    def __init__(
+        self,
+        coordinator: HabiticaDataUpdateCoordinator,
+    ) -> None:
+        """Initialize calendar entity."""
+        super().__init__(coordinator, self.entity_description)
+
+
+class HabiticaTodosCalendarEntity(HabiticaCalendarEntity):
+    """Habitica todos calendar entity."""
+
+    entity_description = CalendarEntityDescription(
+        key=HabiticaCalendar.TODOS,
+        translation_key=HabiticaCalendar.TODOS,
+    )
+
+    def dated_todos(
+        self, start_date: datetime, end_date: datetime | None = None
+    ) -> list[CalendarEvent]:
+        """Get all dated todos."""
+
+        events = []
+        for task in self.coordinator.data.tasks:
+            if (
+                task["type"] == HabiticaTaskType.TODO  # only todos
```

Comment:
OK to omit these comments since they are pretty easy to follow, the code being self explanatory. (Maybe the "due date" one is useful though since its just called "date")